### ISSUE TITLE
StandaloneAhcWSClient does not retry on read timeout

### ISSUE THREAD

**Issue by jsw** [playframework/play-ws#200]
### Play WS Version (2.5.x / etc)

`"com.typesafe.play" %% "play-ahc-ws-standalone" % "1.1.3"`

### API (Scala / Java / Neither / Both)

Scala

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

MacOS Sierra 10.12.6

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. Expect the client to retry on read (idle) timeout

### Actual Behavior

1. Evidence indicates there is no retry, based on the timing numbers printed in the test case below

### Reproducible Test Case

Sample code that demonstrates the issue

```
package example

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import org.scalatest._
import play.api.libs.ws.ahc.{AhcWSClientConfigFactory, StandaloneAhcWSClient}

import scala.concurrent.ExecutionContext.Implicits._
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}
import scala.util.{Failure, Success, Try}

class HelloSpec extends FlatSpec with Matchers {
  "play-ws" should "work" in {
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()

    val defaultConfig = AhcWSClientConfigFactory.forConfig()
    val config = defaultConfig.copy(maxRequestRetry = 6, wsClientConfig = defaultConfig.wsClientConfig.copy(idleTimeout = 2.seconds))
    println(s"config = ${config}")
    val wsClient = StandaloneAhcWSClient(config)
    val start = System.currentTimeMillis()
    val f = call(wsClient)
      .andThen { case _ => wsClient.close() }
      .andThen { case _ => system.terminate() }
    Try {
      Await.result(f, 20.seconds)
    } match {
      case Success(_) => println("Success")
      case Failure(_) =>
        val duration = System.currentTimeMillis() - start
        println(s"Failed in ${duration} ms")
    }
  }

  def call(wsClient: StandaloneAhcWSClient): Future[Unit] = {
    wsClient.url("https://httpbin.org/delay/10").get().map { response =>
      println(s"Got a response ${response.statusText}")
    }
  }
}
```

**Comment by slandelle** [comment]
> Expect the client to retry on read (idle) timeout

Nah. `maxRetry` is used to retry on some socket crashes, such as starting to write before realizing the peer has already closed the socket on its side.
If you want to perform an applicative retry, you must implement it.

**Comment by jsw** [comment]
@slandelle Is there documentation that enumerates the exact set of conditions that retry applies to?

**Comment by slandelle** [comment]
@jsw The intent is to cope with the fact that network is not instant and connection might be closed when trying to write. The typical use case is fetching a keep-alive connection from the pool while the peer has already closed on its side but we haven't received the FIN yet.
Regarding implementation, this is super complex and even I don't really understand it. I inherited this from the original project author, and he had lost track of the implementation details. It typically involves stacktrace scanning :( I suspect some of those cope with old JDK bugs that have been solved since, but some are the only solution, as JDK IO Exception handling is the worst crap ever (untyped IOException with localized messages).

**Comment by jsw** [comment]
@slandelle Thanks for the detailed explanation. I'm still trying to understand at a somewhat higher level the types of conditions that are retried. The initial response to [this post](https://groups.google.com/forum/#!topic/play-framework/0crA-vv0FGY) suggests that it's all `IOException`s, which include connect error, connect timeout, and read timeout, although read timeout is manifesting as a `TimeoutException`, which is not an `IOException`.

Basically, I'm trying to determine what play-ws/AHC/netty is retrying, and what I should take care of retrying in application code. Any guidance you have is much appreciated (either here or in the google group).

**Comment by slandelle** [comment]
I hope to be able to clean this up someday. Clearly, the intent is only retry on connection closed by peer while writing. Maybe add later a second distinct retry mechanism/parameter for general retry.

**Comment by jsw** [comment]
Got it. For reference, could you link the relevant part of the code that determines whether to retry?
